File: src/Json/Process/Factor.fs
namespace ARCtrl.Json


open Thoth.Json.Core

open ARCtrl
open ARCtrl.Process
open System.IO


module Factor =  

    module ISAJson = 

        let genID (f : Factor) = 
            match f.Name with
            | Some name -> $"#Factor/{name}"
            | None -> 
                match f.FactorType with
                | Some factorType -> 
                    $"#Factor/{OntologyAnnotation.ROCrate.genID factorType}"
                | None -> "#EmptyFactor"

        let encoder (idMap : IDTable.IDTableWrite option) (value : Factor) = 
            let f (value : Factor) =
                [
                    Encode.tryInclude "@id" Encode.string (value |> genID |> Some)
                    Encode.tryInclude "factorName" Encode.string value.Name
                    Encode.tryInclude "factorType" (OntologyAnnotation.ISAJson.encoder idMap) value.FactorType
                    Encode.tryIncludeSeq "comments" (Comment.ISAJson.encoder idMap) (value.Comments |> Option.defaultValue (ResizeArray()))
                ]
                |> Encode.choose
                |> Encode.object
            match idMap with
            | None -> f value
            | Some idMap -> IDTable.encode genID f value idMap

        let decoder : Decoder<Factor> =
            Decode.object (fun get ->
                {                    
                    Name = get.Optional.Field "factorName" Decode.string
                    FactorType = get.Optional.Field "factorType" (OntologyAnnotation.ISAJson.decoder)
                    Comments = get.Optional.Field "comments" (Decode.resizeArray (Comment.ISAJson.decoder))
                }
            )

File: js/Json/Process/Factor.js
import { OntologyAnnotation_ISAJson_decoder, OntologyAnnotation_ISAJson_encoder, OntologyAnnotation_ROCrate_genID } from "../OntologyAnnotation.js";
import { ofArray, choose } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { defaultArg, map } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { tryIncludeSeq, tryInclude } from "../Encode.js";
import { ISAJson_decoder, ISAJson_encoder } from "../Comment.js";
import { map as map_1 } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";
import { encode } from "../IDTable.js";
import { resizeArray, string, object } from "../../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { Factor } from "../../Core/Process/Factor.js";

export function genID(f) {
    const matchValue = f.Name;
    if (matchValue == null) {
        const matchValue_1 = f.FactorType;
        if (matchValue_1 == null) {
            return "#EmptyFactor";
        }
        else {
            return `#Factor/${OntologyAnnotation_ROCrate_genID(matchValue_1)}`;
        }
    }
    else {
        return `#Factor/${matchValue}`;
    }
}

export function encoder(idMap, value) {
    const f_1 = (value_1) => {
        const values = choose((tupledArg) => map((v_1) => [tupledArg[0], v_1], tupledArg[1]), ofArray([tryInclude("@id", (value_2) => ({
            Encode(helpers) {
                return helpers.encodeString(value_2);
            },
        }), genID(value_1)), tryInclude("factorName", (value_4) => ({
            Encode(helpers_1) {
                return helpers_1.encodeString(value_4);
            },
        }), value_1.Name), tryInclude("factorType", (oa) => OntologyAnnotation_ISAJson_encoder(idMap, oa), value_1.FactorType), tryIncludeSeq("comments", (comment) => ISAJson_encoder(idMap, comment), defaultArg(value_1.Comments, []))]));
        return {
            Encode(helpers_2) {
                const arg = map_1((tupledArg_1) => [tupledArg_1[0], tupledArg_1[1].Encode(helpers_2)], values);
                return helpers_2.encodeObject(arg);
            },
        };
    };
    if (idMap != null) {
        return encode(genID, f_1, value, idMap);
    }
    else {
        return f_1(value);
    }
}

export const decoder = object((get$) => {
    let objectArg, objectArg_1, arg_5, objectArg_2;
    return new Factor((objectArg = get$.Optional, objectArg.Field("factorName", string)), (objectArg_1 = get$.Optional, objectArg_1.Field("factorType", OntologyAnnotation_ISAJson_decoder)), (arg_5 = resizeArray(ISAJson_decoder), (objectArg_2 = get$.Optional, objectArg_2.Field("comments", arg_5))));
});

